### ISSUE TITLE
AttributeError: 'function' object has no attribute 'expandtabs'

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#932]
I just have this error generating the doc:

```
#!python

$ sphinx-build -b html . _build
Running Sphinx v1.1.3
loading pickled environment... done
building [html]: targets for 1 source files that are out of date
updating environment: 7 added, 1 changed, 0 removed
reading sources... [ 37&#37;] backend                                                                                                                     
Exception occurred:
  File "/usr/local/lib/python2.7/dist-packages/sphinx/util/docstrings.py", line 24, in prepare_docstring
    lines = s.expandtabs().splitlines()
AttributeError: 'function' object has no attribute 'expandtabs'
The full traceback has been saved in /tmp/sphinx-err-o3_3Wv.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
Either send bugs to the mailing list at <http://groups.google.com/group/sphinx-dev/>,
or report them in the tracker at <https://bitbucket.org/birkenfeld/sphinx/issues/>. Thanks!



#!python

$ cat /tmp/sphinx-err-o3_3Wv.log
# Sphinx version: 1.1.3
# Python version: 2.7.2+
# Docutils version: 0.7 release
# Jinja2 version: 2.6
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/sphinx/cmdline.py", line 189, in main
    app.build(force_all, filenames)
  File "/usr/local/lib/python2.7/dist-packages/sphinx/application.py", line 204, in build
    self.builder.build_update()
  File "/usr/local/lib/python2.7/dist-packages/sphinx/builders/__init__.py", line 196, in build_update
    'out of date' &#37; len(to_build))
  File "/usr/local/lib/python2.7/dist-packages/sphinx/builders/__init__.py", line 216, in build
    purple, length):
  File "/usr/local/lib/python2.7/dist-packages/sphinx/builders/__init__.py", line 120, in status_iterator
    for item in iterable:
  File "/usr/local/lib/python2.7/dist-packages/sphinx/environment.py", line 613, in update_generator
    self.read_doc(docname, app=app)
  File "/usr/local/lib/python2.7/dist-packages/sphinx/environment.py", line 761, in read_doc
    pub.publish()
  File "/usr/lib/pymodules/python2.7/docutils/core.py", line 203, in publish
    self.settings)
  File "/usr/lib/pymodules/python2.7/docutils/readers/__init__.py", line 69, in read
    self.parse()
  File "/usr/lib/pymodules/python2.7/docutils/readers/__init__.py", line 75, in parse
    self.parser.parse(self.input, document)
  File "/usr/lib/pymodules/python2.7/docutils/parsers/rst/__init__.py", line 157, in parse
    self.statemachine.run(inputlines, document, inliner=self.inliner)
  File "/usr/lib/pymodules/python2.7/docutils/parsers/rst/states.py", line 170, in run
    input_source=document['source'])
  File "/usr/lib/pymodules/python2.7/docutils/statemachine.py", line 233, in run
    context, state, transitions)
  File "/usr/lib/pymodules/python2.7/docutils/statemachine.py", line 454, in check_line
    return method(match, context, next_state)
  File "/usr/lib/pymodules/python2.7/docutils/parsers/rst/states.py", line 2706, in underline
    self.section(title, source, style, lineno - 1, messages)
  File "/usr/lib/pymodules/python2.7/docutils/parsers/rst/states.py", line 329, in section
    self.new_subsection(title, lineno, messages)
  File "/usr/lib/pymodules/python2.7/docutils/parsers/rst/states.py", line 398, in new_subsection
    node=section_node, match_titles=1)
  File "/usr/lib/pymodules/python2.7/docutils/parsers/rst/states.py", line 284, in nested_parse
    node=node, match_titles=match_titles)
  File "/usr/lib/pymodules/python2.7/docutils/parsers/rst/states.py", line 195, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "/usr/lib/pymodules/python2.7/docutils/statemachine.py", line 233, in run
    context, state, transitions)
  File "/usr/lib/pymodules/python2.7/docutils/statemachine.py", line 454, in check_line
    return method(match, context, next_state)
  File "/usr/lib/pymodules/python2.7/docutils/parsers/rst/states.py", line 2706, in underline
    self.section(title, source, style, lineno - 1, messages)
  File "/usr/lib/pymodules/python2.7/docutils/parsers/rst/states.py", line 329, in section
    self.new_subsection(title, lineno, messages)
  File "/usr/lib/pymodules/python2.7/docutils/parsers/rst/states.py", line 398, in new_subsection
    node=section_node, match_titles=1)
  File "/usr/lib/pymodules/python2.7/docutils/parsers/rst/states.py", line 284, in nested_parse
    node=node, match_titles=match_titles)
  File "/usr/lib/pymodules/python2.7/docutils/parsers/rst/states.py", line 195, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "/usr/lib/pymodules/python2.7/docutils/statemachine.py", line 233, in run
    context, state, transitions)
  File "/usr/lib/pymodules/python2.7/docutils/statemachine.py", line 454, in check_line
    return method(match, context, next_state)
  File "/usr/lib/pymodules/python2.7/docutils/parsers/rst/states.py", line 2281, in explicit_markup
    nodelist, blank_finish = self.explicit_construct(match)
  File "/usr/lib/pymodules/python2.7/docutils/parsers/rst/states.py", line 2293, in explicit_construct
    return method(self, expmatch)
  File "/usr/lib/pymodules/python2.7/docutils/parsers/rst/states.py", line 2035, in directive
    directive_class, match, type_name, option_presets)
  File "/usr/lib/pymodules/python2.7/docutils/parsers/rst/states.py", line 2086, in run_directive
    result = directive_instance.run()
  File "/usr/local/lib/python2.7/dist-packages/sphinx/ext/autodoc.py", line 1303, in run
    documenter.generate(more_content=self.content)
  File "/usr/local/lib/python2.7/dist-packages/sphinx/ext/autodoc.py", line 734, in generate
    self.document_members(all_members)
  File "/usr/local/lib/python2.7/dist-packages/sphinx/ext/autodoc.py", line 659, in document_members
    check_module=members_check_module and not isattr)
  File "/usr/local/lib/python2.7/dist-packages/sphinx/ext/autodoc.py", line 731, in generate
    self.add_content(more_content)
  File "/usr/local/lib/python2.7/dist-packages/sphinx/ext/autodoc.py", line 1046, in add_content
    ModuleLevelDocumenter.add_content(self, more_content)
  File "/usr/local/lib/python2.7/dist-packages/sphinx/ext/autodoc.py", line 471, in add_content
    docstrings = self.get_doc(encoding)
  File "/usr/local/lib/python2.7/dist-packages/sphinx/ext/autodoc.py", line 1034, in get_doc
    doc.append(prepare_docstring(docstring))
  File "/usr/local/lib/python2.7/dist-packages/sphinx/util/docstrings.py", line 24, in prepare_docstring
    lines = s.expandtabs().splitlines()
AttributeError: 'function' object has no attribute 'expandtabs'
```

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/932
- Originally reported by: Ángel Guzmán Maeso
- Originally created at: 2012-05-22T09:54:12.702

**Comment by shimizukawa** [comment]
_From Joshua Skelton on 2013-04-03 05:48:57+00:00_

It looks like the prepare_docstrings method in the docstrings module expects the parameter s to be of type String. In my case it was of type unicode which doesn't have an expandtabs method.

As a work around, I just used the built-in function str() to convert s to the correct type and it appears to be working just fine now.

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2014-01-10 20:51:20+00:00_

Closes #932: autodoc: Do not crash if `__doc__` is not a string.

→ <<cset 38d73fb056bef6dec97e88be7e8383a1242a885a>>
